hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/KeyOutputStream.java

```diff
@@ -555,6 +565,12 @@ public synchronized void close() throws IOException {
       if (!isException) {
         Preconditions.checkArgument(writeOffset == offset);
       }
+      if (atomicKeyCreation) {
+        long expectedSize = blockOutputStreamEntryPool.getDataSize();
+        Preconditions.checkState(expectedSize == offset,
```

Comment:
Since expect this failure mode to occur on a regular basis due to client connection issues, we should not be using pre condition but a simple check and skipping commit key if the entire key has not been written at close.

Reply:
If an error code needs to be returned to the client, we should explicitly check that before closing. Throwing an exception is too heavy-handed for a regular check.

Reply:
Understood, but throwing an exception is currently the most efficient and simple solution. If we don't throw exceptions, we may need to do a lot of extra checking, which would require a lot of code changes.


One possible reason is the S3G uses file semantics rather than key semantics to write keys.
File semantics for writing a key requires 1. create, 2 write, 3 close (including commit), and it is difficult to pass some state between these three separate and distinct steps, whereas with key semantics for put (put=create+write+close), where all the state is managed inside the put, it is simpler.

Reply:
Please file an improvement Jira to not raise exception but refactor the code in a later PR?

Reply:
OK, https://issues.apache.org/jira/browse/HDDS-9640